### ISSUE TITLE
Get Coin Orientation

### ISSUE THREAD

**Issue by pkrush** [GemHunt/CoinSorter#16]
Getting coin orientation would be a major plus to this project that can be built upon. Features such as date & mint mark could then be targeted better for classification. 

This also would be create for manual labeling and a using the system as a generic imaging tool. 

I looked into this briefly. RNN networks are a possibility. 

Just matching sift features on the same coin works:
http://www.mathworks.com/help/vision/ref/matchfeatures.html
This is just not robust enough. 

In the scrips directory there is a MATLAB RimCrop.m script that interpolates the rim into a 120x360 rectangular image. I'm sure 120 is too much, but I left it thick so people can understand where it came from. 

![donut2](https://cloud.githubusercontent.com/assets/13420310/9333282/b155e6fc-458e-11e5-824c-13a712b9e7be.png)

**Comment by pkrush** [comment]
If you are interested in working on this I put over 80,000 labeled penny images out on [GemHunt.com](http://www.gemhunt.com). If you need something different please leave a comment. 

I want to work on this so bad! I would love to try some common sense ideas such as:

**Brute force:**
1. Manually labeling a 1000 or so images
2. Create 360 images for each labeled image by rotating them one for each degree
3. Training this 360,000 image network. 

or

**Building a network based on matching say many 28x28 outer ring segments.**

or 

**I would love to try a cluster the images, just to see what happens:**
        This would be like K-means, but using AlexNet as the distance function
1. Randomly assign 360 categories.
2. Train this data set 
3. Classify the images to the 360 categories using the newly trained network. 
4. Randomly re-fill empty or little filled categories with weak results. 
5. Repeat steps 2-4

The trouble is, from past machine learning I done, **using intuition is not the best way to plan!**
- I need to learn more about building custom networks first. 
- I have to get the hardware better so others can use it. 
- I should ask for help before I try this, which leads back to task 2.

**Comment by pkrush** [comment]
I tried posting a question about this on the Google Caffe Users group, but I think I have not done enough home work for people to care about it. 

**I have a feeling this is the best way:**
- Write software to label Images, and label a 1000 pennies.
- Write a script to crop ring segments, say 10 images 36x36 crops 
- Train a network with 360 small crops. 
- At this point I will know better how to classify. 

**But, I have a stronger feeling I need to learn CNN’s better!**
- Get a local server up because AWS is slowing me down. 
- Understand CNN’s better. (I do think I have done the machine learning basics to go here) 
- Change image sizes in the network. 
- Understand padding & cropping. 
- Fully understand the layers
- Fully understand the deep learning NVidia QwikLabs
- Learn to use caffe without digits.

Sure an RNN might be better, but I really think I need to understand CNNs first to benefit from this. 

By the way, yes, I am basically talking to myself at this point. No one is reading this little corner of GitHub! This is OK; writing this out helps me work through these issues.

**Comment by pkrush** [comment]
Success! This was the key to classifying other features such as date and mint mark. BTW, I did get date and mint mart to work!

**I did this "Brute force" plan with 45% accuracy:**
1.) Labeled 1400 images with 0-3 degree accuracy. 
2.) Rotated each of this 360 times one for each degree
3.) Convert to gray and scaled them 40x40
4.) DIGITS scale them to 28x28
5.) Used LeNet, with 4x on the 3 different weights. 

45% is great because the coins where not within 1% anyway. 

**This is also accurate as well:**
Building a network based on matching say many 28x28 outer ring segments.

So now the plan is to classify with both networks and manually fix issues and retrain.
